packages/kokkos/core/src/impl/Kokkos_Core.cpp

```diff
@@ -93,21 +93,80 @@ setenv("MEMKIND_HBW_NODES", "1", 0);
   const int use_numa = args.num_numa;
 #endif
 #if defined( KOKKOS_ENABLE_CUDA ) || defined( KOKKOS_ENABLE_ROCM )
+  const std::string ctest_device_type = args.ctest_device_type;
   int use_gpu = args.device_id;
   const int ndevices = args.ndevices;
   const int skip_device = args.skip_device;
   // if the exact device is not set, but ndevices was given, assign round-robin using on-node MPI rank
-  if (use_gpu < 0 && ndevices >= 0) {
+  if (use_gpu < 0) {
     auto local_rank_str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK"); //OpenMPI
     if (!local_rank_str) local_rank_str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK"); //MVAPICH2
     if (!local_rank_str) local_rank_str = std::getenv("SLURM_LOCALID"); //SLURM
-    if (local_rank_str) {
+
+    auto ctest_process_count_str = std::getenv("CTEST_PROCESS_COUNT"); //CTest
```

Comment:
For the purposes of experimenting with this, can we define another env var like `KOKKOS_USE_GPU_CTEST_INFO` and if set to `1`, then read in these vars? That would avoid needing to pass `--kokkos-device-ctest=gpus` into every function and it would allow us to test this strategy on the entire Trilinos test suite (not just a few MueLu tests).

Reply:
TriBITS will also need a way to set the `PROCESSES`` property of each test to list the GPU requirements. (The value will also be different for each test, since they have different numbers of processes.) What is your recommended strategy for doing this?

Reply:
@KyleFromKitware,

Seems like the all look like:

```
TRIBITS_ADD_TEST( ...
    NUM_MPI_PROCS <n>
    PROCESSES "<n>,gpus:512"
    ...
    )
```

where `<n>` is just the number of MPI processes/ranks, right?

What is the significance of the `512` in `gpu:512` in all of these?  That was the estimate of memory (which we have no idea what it actually is)?

How do we want CTest to actually limit the number of tests that can run using the GPUs?  Trying to estimate memory requirements is a complete guessing game.

Reply:
Doing it by memory was just a quick example to show how it might work. I think doing it by threads would ultimately be better (no more than X threads on one GPU at a time... we will need to come up with a way to define how many threads one GPU can handle.)

As far as applying this to tests on a global level, I'm thinking we could set a variable like the following:

```cmake
SET(TRIBITS_GPU_THREADS_PER_PROCESS 1)
```

and then `TRIBITS_ADD_TEST()` will set this property appropriately.

Reply:
@KyleFromKitware, I don't think you can control the number of threads run on the GPU, can you?  I think it just uses as many as it wants, no?

Reply:
Memory has the same problem.

Either way, we need a way to define how much capacity a GPU has, and how much of that capacity is consumed by a single process.

If that number is the same (or roughly the same) for every test-process, then we can set a global TriBITS variable. If not, then every test has to have this property individually set.

Reply:
Or if the vast majority uses the same capacity, but there are a few outlier tests, then exceptions could be made for those individual tests.

Reply:
@KyleFromKitware, I think we need to know more about how a GPU managed multiple processes that run kernels on them.  Do the requests from different processes to run kernels on a GPU run at the same time?  Do they run in serial?  What about the RAM used by each process that runs things on the GPU?  We need to understand all of that before we can come with with a good strategy for having ctest limit what runs on the GPUs.

I think we need another meeting with some Kokkos developers to understand these constraints with how things run on a GPU.

Reply:
Agreed.

In the meantime, do we still want to proceed with modifying Kokkos for CTest awareness, or would you prefer the wrapper script strategy that we discussed yesterday?

Reply:
@KyleFromKitware, I think the easiest thing to try is:

* Add an env var to Kokkos to like `KOKKOS_USE_GPU_CTEST_INFO` that if set to `1`, has the same impact as the current command-line argument `--kokkos-device-ctest=gpus`.

* Update TriBITS on this branch to set the env var ``KOKKOS_USE_GPU_CTEST_INFO=1` for each test  (using ctest property `ENVIRONMENT`).

* Update TriBITS on this branch to set the ctest property `PROCESSES "<n>,gpus:512"` where `<n>` is taken from the `NUM_MPI_PROCS <n>` argument for every test.   (Don't need a `PROCESSES` property exposed in `TRIBITS_ADD_TEST()`.)

I think if we do that, then I think this will work and then we can run the entire Trilinos test suite with no modifications to any Trilinos CMakeLists.txt files.

NOTE: This is not quite what we want for production because ctest will limit running tests that may not even require a GPU but that may not be a big issue either.